### ISSUE TITLE
Cannot build mlpack from source after distro upgrade.

### ISSUE THREAD

**Issue by niteya-shah** [mlpack/mlpack#1825]
#### Issue description
I recently updated to Ubuntu 19.04 and when I tried to recompile mlpack from source I started getting this weird error .
#### Your environment

 * version of mlpack: master branch
 * operating system: Ubuntu 19.04
 * compiler: g++ 8.30
 * version of dependencies (Boost/Armadillo): 1.67 / ARMA version: 9.200.7
 * any other environment information you think is relevant: It used to work in previous version without any issues .

#### Steps to reproduce
$ mkdir build && cd build
$ sudo cmake -D PROFILE=ON -D BUILD_PYTHON_BINDINGS=ON -BUILD_TESTS=ON -D FORCE_CXX11=ON -D USE_OPENMP=ON -D TEST_VERBOSE=ON -D MATHJAX_JS_PATH=/home/niteya/MathJax/ -D MATHJAX=ON  ../mlpack
$ make install
(build is outside of mlpack folder)
#### Expected behavior
The Library should be installed .
#### Actual behavior
 cmake output
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- cotire 1.7.10 loaded.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Armadillo: /usr/lib/libarmadillo.so (found suitable version "9.200.7", minimum required is "6.500.0") 
-- Armadillo libraries: /usr/lib/libarmadillo.so
-- Could NOT find ensmallen (missing: ENSMALLEN_INCLUDE_DIR) 
-- [download 0% complete]
-- [download 1% complete]
-- [download 2% complete]
-- [download 3% complete]
-- [download 4% complete]
-- [download 5% complete]
-- [download 6% complete]
-- [download 7% complete]
-- [download 8% complete]
-- [download 9% complete]
-- [download 10% complete]
-- [download 11% complete]
-- [download 12% complete]
-- [download 13% complete]
-- [download 14% complete]
-- [download 15% complete]
-- [download 16% complete]
-- [download 17% complete]
-- [download 18% complete]
-- [download 19% complete]
-- [download 20% complete]
-- [download 21% complete]
-- [download 22% complete]
-- [download 23% complete]
-- [download 24% complete]
-- [download 25% complete]
-- [download 26% complete]
-- [download 27% complete]
-- [download 28% complete]
-- [download 29% complete]
-- [download 30% complete]
-- [download 31% complete]
-- [download 32% complete]
-- [download 33% complete]
-- [download 34% complete]
-- [download 35% complete]
-- [download 36% complete]
-- [download 37% complete]
-- [download 38% complete]
-- [download 39% complete]
-- [download 40% complete]
-- [download 41% complete]
-- [download 42% complete]
-- [download 43% complete]
-- [download 44% complete]
-- [download 45% complete]
-- [download 46% complete]
-- [download 47% complete]
-- [download 48% complete]
-- [download 49% complete]
-- [download 50% complete]
-- [download 51% complete]
-- [download 52% complete]
-- [download 53% complete]
-- [download 54% complete]
-- [download 55% complete]
-- [download 56% complete]
-- [download 57% complete]
-- [download 58% complete]
-- [download 59% complete]
-- [download 60% complete]
-- [download 61% complete]
-- [download 62% complete]
-- [download 63% complete]
-- [download 64% complete]
-- [download 65% complete]
-- [download 66% complete]
-- [download 67% complete]
-- [download 68% complete]
-- [download 69% complete]
-- [download 70% complete]
-- [download 71% complete]
-- [download 72% complete]
-- [download 73% complete]
-- [download 74% complete]
-- [download 75% complete]
-- [download 76% complete]
-- [download 77% complete]
-- [download 78% complete]
-- [download 79% complete]
-- [download 80% complete]
-- [download 81% complete]
-- [download 82% complete]
-- [download 83% complete]
-- [download 84% complete]
-- [download 85% complete]
-- [download 86% complete]
-- [download 87% complete]
-- [download 88% complete]
-- [download 89% complete]
-- [download 90% complete]
-- [download 91% complete]
-- [download 92% complete]
-- [download 93% complete]
-- [download 94% complete]
-- [download 95% complete]
-- [download 96% complete]
-- [download 97% complete]
-- [download 98% complete]
-- [download 99% complete]
-- [download 100% complete]
-- Successfully downloaded ensmallen into /home/niteya/mlpack/build/UILD_TESTS=ON/deps/ensmallen-1.14.2/
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   program_options
--   unit_test_framework
--   serialization
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Found Git: /usr/bin/git (found version "2.20.1") 
-- gitversion.hpp is already up to date.
-- Regenerating arma_config.hpp.
-- Not building Markdown bindings.
-- Found PythonInterp: /usr/bin/python (found version "2.7.16") 
-- Found Python: /usr/bin/python
-- Found PY_distutils: /usr/lib/python2.7/distutils  
-- Found PY_Cython: /usr/lib/python2.7/dist-packages/Cython  
-- Found PY_numpy: /usr/lib/python2.7/dist-packages/numpy  
-- Found PY_pandas: /usr/lib/python2.7/dist-packages/pandas  
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- CXX target mlpack cotired.
-- CXX target mlpack_test cotired.
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 
-- Found MathJax: /home/niteya/MathJax  
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/niteya/mlpack/build/UILD_TESTS=ON

make install output

[  1%] Updating arma_config.hpp (if necessary)
-- Regenerating arma_config.hpp.
[  1%] Built target mlpack_arma_config
[  1%] Updating gitversion.hpp (if necessary)
-- Found Git: /usr/bin/git (found version "2.20.1") 
-- gitversion.hpp is already up to date.
[  1%] Built target mlpack_gitversion
Moving header files to include/mlpack/
c[  1%] Built target mlpack_headers
[  1%] Building CXX precompiled header src/mlpack/cotire/mlpack_CXX_prefix.hxx.gch
In file included from /home/niteya/mlpack/build/src/mlpack/cotire/mlpack_CXX_prefix.cxx:4:0,
                 from /home/niteya/mlpack/build/src/mlpack/cotire/mlpack_CXX_prefix.hxx:4:
/home/niteya/mlpack/mlpack/src/mlpack/prereqs.hpp:74:10: fatal error: boost/serialization/serialization.hpp: No such file or directory
 #include <boost/serialization/serialization.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMake Error at /home/niteya/mlpack/mlpack/CMake/cotire.cmake:1885 (message):
  cotire: error 1 precompiling
  /home/niteya/mlpack/build/src/mlpack/cotire/mlpack_CXX_prefix.hxx.
Call Stack (most recent call first):
  /home/niteya/mlpack/mlpack/CMake/cotire.cmake:3558 (cotire_precompile_prefix_header)

make[2]: *** [src/mlpack/CMakeFiles/mlpack.dir/build.make:64: src/mlpack/cotire/mlpack_CXX_prefix.hxx.gch] Error 1
make[1]: *** [CMakeFiles/Makefile2:492: src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

I have checked and the boost include is there in /usr/include/boost , so I dont know why the compiler cant find the header files. I also tried to manually add them via BOOST_INCLUDEDIR in cmake but still the error persisted.

**Comment by niteya-shah** [comment]
I wrongly wrote the cmake command and changing it fixed the issue :)

**Comment by louisway** [comment]
Hi, I got the same error. Could you please provide which part was wrong?

**Comment by zoq** [comment]
@louisway can you post the commands you used to build mlpack, the ``CMakeCache.txt`` file might be interesting as well.

**Comment by louisway** [comment]
Hi zoq, I think I was using a normal command:
 cmake -D DEBUG=ON -D PROFILE=ON -D USE_OPENMP=ON  -DCMAKE_PREFIX_PATH=~/mlpack  ../

FindBoost showed that it could find the boost:
    Found Boost 1.70.0 at /home/louis/boost/lib/cmake/Boost-1.70.0
     --   Requested configuration: QUIET REQUIRED COMPONENTS 
    program_options;unit_test_framework;serialization
    -- Found boost_headers 1.70.0 at /home/louis/boost/lib/cmake/boost_headers-1.70.0
    -- Found boost_program_options 1.70.0 at 
    /home/louis/boost/lib/cmake/boost_program_options-1.70.0
    --   libboost_program_options.so.1.70.0
    -- Adding boost_program_options dependencies: headers
    -- Found boost_unit_test_framework 1.70.0 at 
    /home/louis/boost/lib/cmake/boost_unit_test_framework-1.70.0
    --   libboost_unit_test_framework.so.1.70.0
    -- Adding boost_unit_test_framework dependencies: headers
    -- Found boost_serialization 1.70.0 at /home/louis/boost/lib/cmake/boost_serialization-1.70.0
    --   libboost_serialization.so.1.70.0
    -- Adding boost_serialization dependencies: headers
    -- Boost 1.49 found.
    -- Found Boost components:
        program_options;unit_test_framework;serialization

I checked the CMakeList.txt file. But it cannot set the Boost_LIBRARY_DIRS, Boost_INCLUDE_DIRS, Boost_LIBRARIES. These variables are empty.
I was doubting if I installed the boost correctly until I saw this post.

**Comment by louisway** [comment]
Hi all, I think I have solved it. Just a hint for others how I solved it: According to https://stackoverflow.com/questions/56036266/boost-libraries-not-defined, the information showed above means FINDBOOST can't locate where the boost is. So the problem becomes easy. just guide your CMakeLists.txt to find the path of boost. What I do is to put list(APPEND CMAKE_PREFIX_PATH "/path/of/boost/") in CMake file. Thank you for your attention.
